Add bookings

rovniipoc opened the pull request:
Добрый день.
Представляю на ревью проект по ТЗ 15го спринта

rovniipoc committed:
Добавить compose.yaml и schema.sql, подготовить к подключению приложения к БД

rovniipoc committed:
Реализовать JPA репозиторий для User

rovniipoc committed:
Реализовать JPA репозиторий для Item и ItemRequest

rovniipoc committed:
Реализовать (частично) JPA репозиторий для Booking

rovniipoc committed:
Реализовать JPA репозиторий для Booking

rovniipoc committed:
Реализовать JPA репозиторий для Comment, реализовать функционал ТЗ 15 спринта, исправить по кодстайлу

rovniipoc committed:
Добавить h2 для тестов в гите

rovniipoc committed:
Добавить application-test.properties с h2 для тестов в гите

rovniipoc committed:
Добавить валидацию дат бронирования

YazykovMikhail commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,67 @@
+    private final BookingService bookingService;
+
+    @PostMapping
+    public Booking create(@RequestHeader("X-Sharer-User-Id") Long userId,
```
Не стоит возвращать Entity 
Возвращай dto, а точнее стоит создать дто для ответа, например 
BookingResponse
а если контракты отличаются в каждом методе, то стоит делать разные дто, например
BookingUpdateResponse
BookingCreateResponse

Это же правило работает и для запросов, если контракты отличаются стит делать разнеы дто, как например
BookingUpdateRequest
BookingCreateRequest
Потому что обычно контракты на создание и обновление отличаются, как минимум для создания не нуэно поле id

Этот комментарий касается всех эндпоинтов и в целом разработки

YazykovMikhail commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,67 @@
+    }
+
+    @PatchMapping("/{bookingId}")
+    public Booking updateBookingStatus(@RequestHeader("X-Sharer-User-Id") Long userId,
```
Лучше вынести в константу 
"X-Sharer-User-Id"
что бы везде использовать эту переменую и не ловить ошибок из за человеческого фактора и не заниматься копипастом

YazykovMikhail commented on src/main/java/ru/practicum/shareit/booking/BookingService.java:
```diff
@@ -0,0 +1,165 @@
+        }
+    }
+
+    public Booking getById(Long bookingId, Long userId) {
```
Лучше сначала проверить что пользователь с таким id существует, а потом уже получать бронирование и проверять разрешение

rovniipoc committed:
Исправить по замечаниям 1го ревью

Изменения:
- заголовок с ID пользователя вынесен в константу в контроллерах;
- доработан DTO: для всех сущностей созданы input и output dto-классы, контроллеры доработаны в этой части;
- добавлена упущенная в getById() BookingService проверка на существование пользователя.
